### PR TITLE
added dat.json url parser to clone command.

### PR DESCRIPTION
**Pull request by TimothyStiles**
I've added dat.json url parser to the clone command as described in #639. `dat clone .`, `dat clone dat.json`, and `dat clone ./path/to/dat.json ./dir` should work as expected.

`dat clone ./path/to/dat.json` without a target directory specified creates a dat repo clone in the current working directory named after the key link that was parsed from dat.json.

`dat clone` does not yet work. It seems to have something to do with [src/parse-args.js](https://github.com/datproject/dat/blob/master/src/parse-args.js). I'll submit an issue about it.

### CODE REVIEW

File: src/commands/clone.js [hunk 1]
```diff
@@ -41,9 +41,23 @@ function clone (opts) {
...
   var parseArgs = require('../parse-args')
   var debug = require('debug')('dat')
-
   var parsed = parseArgs(opts)
+  var path = require('path')
+
   opts.key = parsed.key || opts._[0] // pass other links to resolver
+
+  // Resolves dat.json path and parses dat url from it.
+  // Null doesn't work yet. May have something to do with ../parse-args. Will submit issue.
```
Comment (ralphtheninja) [inline src/commands/clone.js]: What does "doesn't work yet" mean?

**Comment by TimothyStiles** [comment]
@ralphtheninja when I added aafb1c3 I was hoping someone would be able to clue me into why `dat clone` wouldn't work without any arguments. It turns out that I had accidentally used `null` instead of `undefined` which I fixed in 6e776c9 and dfa61dd. Everything should now work as specified in #639 and [publish.md](https://github.com/datproject/docs/blob/master/docs/cookbook/publish.md) minus `dat .`.

I made another change in 6e776c9 and dfa61dd so now if an output directory isn't specified when a url is parsed from a dat.json file the working directory becomes the output directory.

There are a couple of caveats that I realized could be fixed that I'm working on now:

- `dat clone ./path/to/output` won't work without catching that the key link is missing.

- `dat clone ./path/to/input ./path/to/output` where `./path/to/input` doesn't include `dat.json` won't work without catching that `dat.json` is missing.

After that I'll try writing and adding a new `'clone - dat.json'` test to [test/clone.js](https://github.com/datproject/dat/blob/master/test/clone.js). I tried writing a test earlier using `'clone - default opts'` as a template. I don't quit understand what's going on with `help.shareFixtures` yet but I'll try to figure it out.

**Comment by TimothyStiles** [comment]
I've fixed the caveats but there still needs to be tests for `dat-is-link` and `clone - dat.json`.

**Comment by ralphtheninja** [comment]
It would be really nice if you could take the code for `dat-is-link` and keep it inside `dat` to start with (it's just a function anyway). It's difficult to read the PR when there's a new module involved. Could put the tests locally to `dat` as well and once everything works, then refactor.

**Comment by ralphtheninja** [comment]
I can write some tests for `dat-is-link` if you want.

**Comment by TimothyStiles** [comment]
> It would be really nice if you could take the code for dat-is-link and keep it inside dat to start with (it's just a function anyway). 

@ralphtheninja, got it. I thought it could be a nice utility to have. Should I take the contents of `dat-is-link` and declare a function such as `isDatLink` within the scope of `clone` before the dat.json parser I added or is there somewhere in `src` to declare helper functions?

Any tests you could write for `is-dat-link` would be great!

**Comment by joehand** [comment]
Thanks for getting started on this. A few comments to add to @ralphtheninja:

* `dat clone` without arguments seems a bit too magical for me. Requiring the third argument will help us make sure it's doing what is expected for now (even if its just `dat clone .`). We can drop that requirement later if needed.
* We have [dat-encoding](https://github.com/juliangruber/dat-encoding) for parsing dat links. Since it is already a dependency, we should reuse this if possible (we can PR any needed changes there).

Since its fairly easy to check if a dat.json exists, its not vital we catch 100% of the dat links. `dat clone datproject.org/joehand/cli-demo` is a valid link but hard to parse. That could refer to either a directory with a dat.json or a link we need to resolve. We should try finding the `dat.json` first and if that fails, then try to resolve the link.

File: src/commands/clone.js [hunk 2]
```diff
@@ -41,10 +41,34 @@ function clone (opts) {
   var onExit = require('../lib/exit')
   var parseArgs = require('../parse-args')
   var debug = require('debug')('dat')
-
   var parsed = parseArgs(opts)
+  var path = require('path')
```
Comment (joehand) [inline src/commands/clone.js]: Can you add the node core modules as the top requires (in this case, right under `require('fs')`?

File: src/commands/clone.js [hunk 3]
```diff
@@ -41,10 +41,34 @@ function clone (opts) {
...
+      if (fs.lstatSync(opts.key).isDirectory() && opts.dir === undefined && fs.existsSync('dat.json')) {
+        datPath = path.resolve('dat.json')
+        opts.dir = opts.key
+      } else if (fs.existsSync(path.resolve(opts.key + '/dat.json'))) {
+        datPath = path.resolve(opts.key + 'dat.son')
+      } else if (path.basename(opts.key) === 'dat.json') {
+        datPath = path.resolve(opts.key)
+      }
+    } else {
+      return console.error("Could not resolve link from 'dat.json'.")
```
Comment (joehand) [inline src/commands/clone.js]: We should handle this error inside the logging, if possible.

File: src/commands/clone.js (hunk 2 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: Will be in next push.

File: src/commands/clone.js [hunk 4]
```diff
@@ -41,10 +41,34 @@ function clone (opts) {
...
+  if (!isDatLink(opts.key)) {
+    var datPath
+    if (opts.key === '.' || opts.key === undefined) { // handles paths for 'dat clone' & 'dat clone .'
+      datPath = path.resolve('dat.json')
+    } else if (fs.existsSync(opts.key)) {
+      if (fs.lstatSync(opts.key).isDirectory() && opts.dir === undefined && fs.existsSync('dat.json')) {
+        datPath = path.resolve('dat.json')
+        opts.dir = opts.key
+      } else if (fs.existsSync(path.resolve(opts.key + '/dat.json'))) {
+        datPath = path.resolve(opts.key + 'dat.son')
```
Comment (TimothyStiles) [inline src/commands/clone.js]: Will be fixed with next push.

File: src/commands/clone.js (hunk 3 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: How would that work? Is there an example you could show me?

Comment (karissa) [inline src/commands/clone.js]: You could do `return bus.emit('exit:error', msg)`

**Comment by TimothyStiles** [comment]
Strange to me. Moving `var path = require('path')` from bottom of deps list to top in c38cc68 caused tests to fail when running on `node.js:6`, but passes on `node.js:4` and `node.js:8` . When I moved it back and ran locally on `node.js:7` all tests passed.

**Comment by joehand** [comment]
We have some inconsistent failing tests because of a hypercore bug, we can rerun it again =). As long as they don't *all* fail it should be okay.

**Comment by ralphtheninja** [comment]
> We should try finding the dat.json first and if that fails, then try to resolve the link.

:+1: 

This should make this a lot easier and more clean.

**Comment by TimothyStiles** [comment]
Am I using `neatLog` as intended in 8d744a6? It does work. A test still needs to be written but is there anything else that should be done or changed?

**Comment by joehand** [comment]
> Also, am I using neatLog as intended in 8d744a6? It does work.

Oops, sorry may have made that more complex than intended. I was thinking your logic could go inside of the existing neat log function, replacing [the opts.key check](https://github.com/TimothyStiles/dat/blob/ee34675faabddf52b88c4f07c974e6dba6ec8392/src/commands/clone.js#L87). Neat that it works with how you did it but don't think we want to init the logger twice.

File: src/commands/clone.js [hunk 5]
```diff
@@ -41,10 +42,37 @@ function clone (opts) {
...
   var debug = require('debug')('dat')
-
   var parsed = parseArgs(opts)
+
   opts.key = parsed.key || opts._[0] // pass other links to resolver
   opts.dir = parsed.dir
+
+  // Resolves dat.json path and parses dat url from it.
+  if (fs.existsSync(opts.key)) {
+    var datPath
```
Comment (joehand) [inline src/commands/clone.js]: Can we set this to `opts.key` here and use it throughout to make it a bit easier to read?

```js
var datPath = opts.key
...
if (path.basename(datPath)) {} 
```

File: src/commands/clone.js [hunk 6]
```diff
@@ -41,10 +42,37 @@ function clone (opts) {
...
+    var neatPath = neatLog(archiveUI, { logspeed: opts.logspeed, quiet: opts.quiet })
+    neatPath.use(trackArchive)
+    neatPath.use(discoveryExit)
+    neatPath.use(onExit)
+    if (path.basename(opts.key) === 'dat.json') { // handles paths for 'dat clone' & 'dat clone .'
+      datPath = path.resolve(opts.key)
+    } else if (fs.lstatSync(opts.key).isDirectory()) {
+      datPath = path.resolve(opts.key + '/dat.json')
```
Comment (joehand) [inline src/commands/clone.js]: Do we need `path.join()` here for windows? (Not 100% if this will resolve correctly with the `/` path separator.

File: src/commands/clone.js [hunk 7]
```diff
@@ -41,10 +42,37 @@ function clone (opts) {
...
+    } else {
+      neatPath.use(function (state, bus) {
+        return bus.emit('exit:error', 'Could not resolve link from dat.json on path: ' + path.resolve(opts.key))
```
Comment (joehand) [inline src/commands/clone.js]: Instead of exiting, we should now assume this is a key (not directory) and try to resolve it, like in the case of `datproject.org/joehand/cli-demo`.

File: src/commands/clone.js (hunk 6 above)
Comment (karissa) [inline src/commands/clone.js]: yeah `path.join`

File: src/commands/clone.js [hunk 8]
```diff
@@ -41,10 +42,37 @@ function clone (opts) {
...
+      })
+    }
+    if (!fs.existsSync(datPath)) {
```
Comment (joehand) [inline src/commands/clone.js]: What is the situation where `fs.existsSync(opts.key)`, above, is true but this is false?

File: src/commands/clone.js [hunk 9]
```diff
@@ -41,10 +42,37 @@ function clone (opts) {
...
+      neatPath.use(function (state, bus) {
+        return bus.emit('exit:error', 'Could not resolve link from dat.json on path: ' + path.resolve(opts.key))
+      })
+    }
+    if (parsed.keyToDirSwitch) opts.dir = undefined
+    opts.dir = opts.dir || path.resolve('.') // if needed changes output directory from undefined to current.
```
Comment (joehand) [inline src/commands/clone.js]: I don't quite understand this. It seems like we'll end up in the wrong directory whenever you specify the directory, e.g. `dat clone /path/to/dir` where `dir/dat.json` exists. Is there a reason we can't use:

```js
opts.dir = opts.dir || datPath
```

File: src/commands/clone.js (hunk 8 above)
Comment (joehand) [inline src/commands/clone.js]: Ah I see, its checking if the `dat.json` exists, not just the path. Same [comment as above](https://github.com/datproject/dat/pull/845/files#r132571106), we should assume this is a key and fail later.

File: src/commands/clone.js (hunk 7 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: By default if `opts.key` isn't a file or directory on the system as determined by `fs.existsSync(opts.key)` then it won't try to resolve it as a directory or file and will try to resolve it as a key like clone normally would before this change. 

```
if (fs.existsSync(opts.key)) {
  var datPath
  var neatPath = neatLog(archiveUI, { logspeed: opts.logspeed, quiet: opts.quiet })
  ...
  ...
```

File: src/commands/clone.js (hunk 9 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: We can't because `parse-args.js` by default set `opts.dir = opts.key` if no `opts.dir` is supplied to `parse-args.js`. When I tried changing this behavior in `parse-args.js` it broke a test (can't remember which) so I just made a flag in the `opts` object that was set to true if only one argument was provided. Without this `dat clone /path/to/dir` would actually run as `dat clone /path/to/dir /path/to/dir`

File: src/commands/clone.js (hunk 7 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: If it gets to this point it's neither a key or directory.

File: src/commands/clone.js (hunk 8 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: Same [comment as above](https://github.com/datproject/dat/pull/845/files#r132571106).

File: src/commands/clone.js (hunk 6 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: Will change in next commit.

File: src/commands/clone.js (hunk 5 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: Will change in next commit.

**Comment by TimothyStiles** [comment]
I've cleaned up the logic a bit. Now errors error when link is actually resolved.

**Comment by joehand** [comment]
> I've cleaned up the logic a bit. Now errors error when link is actually resolved.

Awesome! This reads a lot nicer now. There are a few small things but I think we can add the tests/docs then get this merged! Let me know if you want help with either of those. Tests are still a bit of a pain right now.

File: src/commands/clone.js [hunk 10]
```diff
@@ -41,10 +42,36 @@ function clone (opts) {
...
+  if (onSystem) {
+    isDir = fs.lstatSync(datPath).isDirectory()
+    isDatJson = (fs.lstatSync(datPath).isFile() && (path.basename(datPath) === 'dat.json'))
+  }
+
+  // if file is a dat.json resolve absolute path. else if directory try to resolve path to dat.json.
+  if (isDatJson) {
+    datPath = path.resolve(datPath)
+  } else if (isDir) {
+    datPath = path.join(datPath + '/dat.json')
```
Comment (joehand) [inline src/commands/clone.js]: `path.join(datPath, 'dat.json')` (the `/` will not give the expected result in windows, so `path.join` replaces it with whatever is correct for the OS).

Comment (TimothyStiles) [inline src/commands/clone.js]: Will fix in next commit.

**Comment by TimothyStiles** [comment]
@joehand help with the tests would be awesome. I should be fine with the docs if it's just updating [publish.md](https://github.com/datproject/docs/blob/master/docs/cookbook/publish.md).

File: src/commands/clone.js (hunk 10 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: I tried fixing in 0c3c3a8 and it ended up breaking behavior when running `dat clone .`. Got the message `Could not resolve link.` Works when changed back in d26dcc1. @joehand any thoughts?

Comment (joehand) [inline src/commands/clone.js]: I'll look in a bit, but it'll be good to throw in some debug statements before/after this block anyways:

```js
debug('cloning from dat.json', datPath) // or something like this
```

You can then run it with `DEBUG=dat dat clone .` to see output.

**Comment by TimothyStiles** [comment]
@joehand tests pass locally but failed on Travis. Is there any way that I can restart Travis tests or can only you do that?

**Comment by joehand** [comment]
> @joehand tests pass locally but failed on Travis. Is there any way that I can restart Travis tests or can only you do that?

Maybe only I can? Not sure if you see the restart arrow on the right side here: https://travis-ci.org/datproject/dat/builds/263573873

Anyways, restarted that one. As long as they pass locally (or some pass) we don't need to worry about travis until we're about to merge.

**Comment by joehand** [comment]
Hey @TimothyStiles see below for some tests (I tried to push but was rejected, did [you allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)? maybe messed up my git-fu). 

The tests where I don't set `cwd` fail because of [the different default expectations](https://github.com/datproject/dat/pull/845#discussion_r132729979).

Tests:

<details>

```js

test('clone - specify directory containing dat.json', function (t) {
  help.shareFixtures(function (_, shareDat) {
    tempDir(function (_, dir, cleanup) {
      fs.writeFileSync(path.join(dir, 'dat.json'), JSON.stringify({url: shareDat.key.toString('hex')}), 'utf8')

      // dat clone /dir
      var cmd = dat + ' clone ' + dir
      var st = spawn(t, cmd)
      var datDir = dir

      st.stdout.match(function (output) {
        var downloadFinished = output.indexOf('Exiting') > -1
        if (!downloadFinished) return false

        var fileList = help.fileList(datDir).join(' ')
        var hasCsvFile = fileList.indexOf('all_hour.csv') > -1
        t.ok(hasCsvFile, 'csv file downloaded')
        var hasDatFolder = fileList.indexOf('.dat') > -1
        t.ok(hasDatFolder, '.dat folder created')
        var hasSubDir = fileList.indexOf('folder') > -1
        t.ok(hasSubDir, 'folder created')
        var hasNestedDir = fileList.indexOf('nested') > -1
        t.ok(hasNestedDir, 'nested folder created')
        var hasHelloFile = fileList.indexOf('hello.txt') > -1
        t.ok(hasHelloFile, 'hello.txt file downloaded')

        st.kill()
        return true
      })
      st.succeeds('exits after finishing download')
      st.stderr.empty()
      st.end(function () {
        cleanup()
        shareDat.close()
      })
    })
  })
})

test('clone - specify directory containing dat.json with cwd', function (t) {
  help.shareFixtures(function (_, shareDat) {
    tempDir(function (_, dir, cleanup) {
      fs.writeFileSync(path.join(dir, 'dat.json'), JSON.stringify({url: shareDat.key.toString('hex')}), 'utf8')

      // cd dir && dat clone /dir/dat.json
      var cmd = dat + ' clone ' + dir
      var st = spawn(t, cmd, {cwd: dir})
      var datDir = dir

      st.stdout.match(function (output) {
        var downloadFinished = output.indexOf('Exiting') > -1
        if (!downloadFinished) return false

        var fileList = help.fileList(datDir).join(' ')
        var hasCsvFile = fileList.indexOf('all_hour.csv') > -1
        t.ok(hasCsvFile, 'csv file downloaded')
        var hasDatFolder = fileList.indexOf('.dat') > -1
        t.ok(hasDatFolder, '.dat folder created')
        var hasSubDir = fileList.indexOf('folder') > -1
        t.ok(hasSubDir, 'folder created')
        var hasNestedDir = fileList.indexOf('nested') > -1
        t.ok(hasNestedDir, 'nested folder created')
        var hasHelloFile = fileList.indexOf('hello.txt') > -1
        t.ok(hasHelloFile, 'hello.txt file downloaded')

        st.kill()
        return true
      })
      st.succeeds('exits after finishing download')
      st.stderr.empty()
      st.end(function () {
        cleanup()
        shareDat.close()
      })
    })
  })
})

test('clone - specify dat.json path', function (t) {
  help.shareFixtures(function (_, shareDat) {
    tempDir(function (_, dir, cleanup) {
      var datJsonPath = path.join(dir, 'dat.json')
      fs.writeFileSync(datJsonPath, JSON.stringify({url: shareDat.key.toString('hex')}), 'utf8')

      // dat clone /dir/dat.json
      var cmd = dat + ' clone ' + datJsonPath
      var st = spawn(t, cmd)
      var datDir = dir

      st.stdout.match(function (output) {
        var downloadFinished = output.indexOf('Exiting') > -1
        if (!downloadFinished) return false

        var fileList = help.fileList(datDir).join(' ')
        var hasCsvFile = fileList.indexOf('all_hour.csv') > -1
        t.ok(hasCsvFile, 'csv file downloaded')
        var hasDatFolder = fileList.indexOf('.dat') > -1
        t.ok(hasDatFolder, '.dat folder created')
        var hasSubDir = fileList.indexOf('folder') > -1
        t.ok(hasSubDir, 'folder created')
        var hasNestedDir = fileList.indexOf('nested') > -1
        t.ok(hasNestedDir, 'nested folder created')
        var hasHelloFile = fileList.indexOf('hello.txt') > -1
        t.ok(hasHelloFile, 'hello.txt file downloaded')

        st.kill()
        return true
      })
      st.succeeds('exits after finishing download')
      st.stderr.empty()
      st.end(function () {
        cleanup()
        shareDat.close()
      })
    })
  })
})

test('clone - specify dat.json path with cwd', function (t) {
  help.shareFixtures(function (_, shareDat) {
    tempDir(function (_, dir, cleanup) {
      var datJsonPath = path.join(dir, 'dat.json')
      fs.writeFileSync(datJsonPath, JSON.stringify({url: shareDat.key.toString('hex')}), 'utf8')

      // cd /dir && dat clone /dir/dat.json
      var cmd = dat + ' clone ' + datJsonPath
      var st = spawn(t, cmd, {cwd: dir})
      var datDir = dir

      st.stdout.match(function (output) {
        var downloadFinished = output.indexOf('Exiting') > -1
        if (!downloadFinished) return false

        var fileList = help.fileList(datDir).join(' ')
        var hasCsvFile = fileList.indexOf('all_hour.csv') > -1
        t.ok(hasCsvFile, 'csv file downloaded')
        var hasDatFolder = fileList.indexOf('.dat') > -1
        t.ok(hasDatFolder, '.dat folder created')
        var hasSubDir = fileList.indexOf('folder') > -1
        t.ok(hasSubDir, 'folder created')
        var hasNestedDir = fileList.indexOf('nested') > -1
        t.ok(hasNestedDir, 'nested folder created')
        var hasHelloFile = fileList.indexOf('hello.txt') > -1
        t.ok(hasHelloFile, 'hello.txt file downloaded')

        st.kill()
        return true
      })
      st.succeeds('exits after finishing download')
      st.stderr.empty()
      st.end(function () {
        cleanup()
        shareDat.close()
      })
    })
  })
})

test('clone - specify dat.json + directory', function (t) {
  help.shareFixtures(function (_, shareDat) {
    tempDir(function (_, dir, cleanup) {
      var datDir = path.join(dir, 'clone-dest')
      var datJsonPath = path.join(dir, 'dat.json') // make dat.json in different dir

      fs.mkdirSync('clone-dest')
      fs.writeFileSync(datJsonPath, JSON.stringify({url: shareDat.key.toString('hex')}), 'utf8')

      // dat clone /dir/dat.json /dir/clone-dest
      var cmd = dat + ' clone ' + datJsonPath + ' ' + datDir
      var st = spawn(t, cmd)

      st.stdout.match(function (output) {
        var downloadFinished = output.indexOf('Exiting') > -1
        if (!downloadFinished) return false

        var fileList = help.fileList(datDir).join(' ')
        var hasCsvFile = fileList.indexOf('all_hour.csv') > -1
        t.ok(hasCsvFile, 'csv file downloaded')
        var hasDatFolder = fileList.indexOf('.dat') > -1
        t.ok(hasDatFolder, '.dat folder created')
        var hasSubDir = fileList.indexOf('folder') > -1
        t.ok(hasSubDir, 'folder created')
        var hasNestedDir = fileList.indexOf('nested') > -1
        t.ok(hasNestedDir, 'nested folder created')
        var hasHelloFile = fileList.indexOf('hello.txt') > -1
        t.ok(hasHelloFile, 'hello.txt file downloaded')

        st.kill()
        return true
      })
      st.succeeds('exits after finishing download')
      st.stderr.empty()
      st.end(function () {
        cleanup()
        shareDat.close()
      })
    })
  })
})
```

</details>

**Comment by joehand** [comment]
Darn, did not mean to merge that.

**Comment by ralphtheninja** [comment]
@joehand Can you restore the PR?

**Comment by joehand** [comment]
> @joehand Can you restore the PR?

Not easily b/c it wasn't merged in github. Have to manually revert. @TimothyStiles you may need to reopen the PR in a bit, sorry.

**Comment by joehand** [comment]
@TimothyStiles please open the PR again, sorry! If you allow edits from maintainers, it'll be easier for me to add tests and what not. Thanks.

**Comment by TimothyStiles** [comment]
@joehand just reopened PR with edits from maintainers. #847

File: src/commands/clone.js (hunk 9 above)
Comment (TimothyStiles) [inline src/commands/clone.js]: Good point. I'll correct this now. Actually will make everything way easier I think.
